### ISSUE TITLE
Complex number autograd inconsistency

### ISSUE THREAD

**Issue by rjkilpatrick** [pytorch/pytorch#47200]
## 🐛 Bug

For any complex number `z = a + bi`, you can consider the imaginary component `b` as `1/2 (z - z*)`. Taking the autograd on the imaginary part, and calculating the imaginary part gives inconsistent results.
If this is intended behaviour, then the documentation should reflect this.

## To Reproduce

``` python
import torch
print(torch.__version__)
# 1.8.0.dev20201101

z = torch.ones(1, requires_grad=True, dtype=torch.complex128)
out = 0.5 * (z - z.conj())
out.backward()
print(z.grad)
# tensor([0.+0.j], dtype=torch.complex128)

z = torch.ones(1, requires_grad=True, dtype=torch.complex128)
out = z.imag
out.backward()
print(z.grad)
# tensor([0.+1.j], dtype=torch.complex128)
```

## Expected behaviour

The autograd of the two methods of determining the imaginary component of `z`, should be consistent.

## Environment
PyTorch version: 1.8.0.dev20201101
Is debug build: True
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Manjaro Linux (x86_64)
GCC version: (GCC) 10.2.0
Clang version: 10.0.1
CMake version: version 3.18.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 970
Nvidia driver version: 450.80.02
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.0.2
/usr/lib/libcudnn_adv_infer.so.8.0.2
/usr/lib/libcudnn_adv_train.so.8.0.2
/usr/lib/libcudnn_cnn_infer.so.8.0.2
/usr/lib/libcudnn_cnn_train.so.8.0.2
/usr/lib/libcudnn_ops_infer.so.8.0.2
/usr/lib/libcudnn_ops_train.so.8.0.2
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.0.dev20201101
[pip3] torchvision==0.9.0.dev20201101
[pip3] torchviz==0.0.1
[conda] blas 1.0 mkl
[conda] cudatoolkit 10.2.89 hfd86e86_1
[conda] mkl 2020.2 256
[conda] mkl-service 2.3.0 py38he904b0f_0
[conda] mkl_fft 1.2.0 py38h23d657b_0
[conda] mkl_random 1.1.1 py38h0573a6f_0
[conda] numpy 1.19.2 py38h54aff64_0
[conda] numpy-base 1.19.2 py38hfa32c7d_0
[conda] torch 1.8.0.dev20201026 pypi_0 pypi
[conda] torchvision 0.9.0.dev20201027 pypi_0 pypi
[conda] torchviz 0.0.1 pypi_0 pypi

cc @ezyang @anjali411 @dylanbespalko @mruberry

**Comment by mruberry** [comment]
Thanks for reporting this issue, @rjkilpatrick. I'll let our complex experts comment.

**Comment by anjali411** [comment]
Hi @rjkilpatrick thanks for filing this issue. This is expected because the formula used to compute complex gradients only guarantees meaningful gradients for real valued objective. So, if I modify your example a little to make the final output real valued, I get the same gradients in both cases.

```
>>> import torch
>>> z = torch.ones(1, requires_grad=True, dtype=torch.complex128)
>>> out = 0.5 * (z - z.conj()).abs()
>>> out.backward()
>>> print(z.grad)
tensor([0.+0.j], dtype=torch.complex128)


>>> z = torch.ones(1, requires_grad=True, dtype=torch.complex128)
>>> out = z.imag.abs()
>>> out.backward()
>>> print(z.grad)
tensor([0.+0.j], dtype=torch.complex128)
```
